domain/resource/state/resource.go

```diff
@@ -259,6 +260,74 @@ AND application_uuid = $resourceIdentity.application_uuid
 	return coreresource.UUID(resource.UUID), nil
 }
 
+// GetResourceUUIDByAppAndResourceName returns the ID of the application
+// resource specified by natural key of application and resource name.
+//
+// The following error types can be expected to be returned:
+//   - [applicationerrors.ApplicationNotFound] is returned if the application is
+//     not found.
+//   - [resourceerrors.ResourceNotFound] if no resource with name exists for
+//     given application.
+func (st *State) GetResourceUUIDByAppAndResourceName(
+	ctx context.Context,
+	appName string,
+	resName string,
+) (coreresource.UUID, error) {
+	db, err := st.DB()
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	// Define the resource identity based on the provided application ID and
+	// name.
+	names := resourceAndAppName{
+		ApplicationName: appName,
+		ResourceName:    resName,
+	}
+	uuid := resourceUUID{}
+
+	// Prepare the SQL statement to retrieve the resource UUID.
+	stmt, err := st.Prepare(`
+SELECT r.uuid AS &resourceUUID.uuid
+FROM   resource AS r
+JOIN   application_resource ar ON r.uuid = ar.resource_uuid
+JOIN   application a           ON ar.application_uuid = a.uuid
+WHERE  r.charm_resource_name = $resourceAndAppName.resource_name 
+AND    a.name = $resourceAndAppName.application_name
```

Comment:
We should use application uuid, it's entirely possible in the future then application name could change. We don't want to tie things to the name

Reply:
I added this whole method so that we could get the resource via the application name rather than the UUID. 

My thinking was that it is a bit heavy to pull in the applications service to the resource handler just for the method to get the application UUID, but I'm happy to remove this new method and just do that.

Reply:
I don't mind adding this, my problem is you should do these queries with the UUID. I think if you returned the UUID when checking the application exists, then I'm happy for this to move forward.

Reply:
Changed to do that.